Core/Graphics/Source/GraphicsImpl.h

```diff
@@ -49,20 +51,20 @@ namespace Babylon
             }
 
         private:
-            friend Impl;
+            friend GraphicsImpl;
 
             arcana::manual_dispatcher<128> m_dispatcher;
         };
 
-        Impl();
-        ~Impl();
-
-        void* GetNativeWindow();
-        void SetNativeWindow(void* nativeWindowPtr, void* windowTypePtr);
+        GraphicsImpl();
+        virtual ~GraphicsImpl();
+        template<typename WindowT>
+        WindowT GetNativeWindow();
```

Comment:
If I understand correctly, this is the change that tries to eliminate the need to compile with the objective C++ flag, is that right? It seems very awkward though, and like we have a weird mix of known and sort of unknown types in the API.

Reply:
My understanding is that if we want this function as a member of `GraphicsImpl` we either A) go back to using void*, B) use the platform type in this header, causing compilation issues with our consumers if the headers arent c++, or c) use some templating to obscure the type in the header.

The alternative is to make it some odd utility function that uses GraphicsImpl? C++ makes these types of situations difficult.

Reply:
I agree that it's a little awkward, but (in my opinion) much less so than having a bunch of C++ libraries that have to be compiled as Objective C++ just so that they can find a type that they don't actually use in C++. The reason for templating this instead of forward-declaring `MTKView` is because Apple defines `MTKView` as an interface, which didn't seem to play nicely with our attempts to forward-declare it in C++. Because this is an internal contract, not a proper outward-facing API, I don't think it's too big a deal for it to have this small degree of awkwardness.

If it's bothersome, though, I can think of two other option to do this cleanly, though the first is (in my opinion) even more awkward: have a special platform-specific struct defined just like the config struct, the only purpose of which is to contain an `MTKView*`, then forward-declare _that_ and use it as the return type. This is awkward because it entails having an entire type which has to be implemented per-platform just for the purpose of avoiding a templated function.

The second approach would be to simply move this function definition to a platform-specific header, using our CMake practices to deliver the right header to the right consumer. This is probably my favorite of the three, but it's still a little awkward because you'd need to take it out of the class, making it `GetNativeWindowFromGraphicsImpl(GraphicsImpl&)` or something, which could get odd. Bottom line, I think all three of these approaches are equally viable and equally awkward, so I don't really have a strong personal preference among them.